lib/auth/webauthncli/u2f.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2021 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webauthncli
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/flynn/u2f/u2fhid"
+	"github.com/flynn/u2f/u2ftoken"
+	"github.com/gravitational/trace"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// DevicePollInterval is the interval between polling attempts on Webauthn or
+// U2F devices.
+// Used by otherwise tight loops such as RunOnU2FDevices and related methods,
+// like Login.
+var DevicePollInterval = 200 * time.Millisecond
+
+var errKeyMissingOrNotVerified = errors.New("key missing or user presence not verified")
+
+// Token represents the actions possible using an U2F/CTAP1 token.
+type Token interface {
+	CheckAuthenticate(req u2ftoken.AuthenticateRequest) error
+	Authenticate(req u2ftoken.AuthenticateRequest) (*u2ftoken.AuthenticateResponse, error)
+	Register(req u2ftoken.RegisterRequest) ([]byte, error)
+}
+
+// u2fDevices, u2fOpen and u2fNewToken allows tests to fake interactions with
+// devices.
+var u2fDevices = u2fhid.Devices
+var u2fOpen = u2fhid.Open
+var u2fNewToken = func(d u2ftoken.Device) Token {
+	return u2ftoken.NewToken(d)
+}
+
+type unknownHandleKey struct {
+	Callback int
+	Path     string
+}
+
+// RunOnU2FDevices polls for new U2F/CTAP1 devices and invokes the callbacks
+// against them in regular intervals, running until either one callback succeeds
+// or the context is cancelled.
+// Typically, each callback represents a {credential,rpid} pair to check against
+// the device.
+// Calling this method using a context without a cancel or deadline means it
+// will execute until successful (which may be never).
+// Most callers should prefer higher-abstraction functions such as Login.
+func RunOnU2FDevices(ctx context.Context, runCredentials ...func(Token) error) error {
+	ticker := time.NewTicker(DevicePollInterval)
+	defer ticker.Stop()
+
+	unknownHandles := make(map[unknownHandleKey]bool)
+	for {
+		switch err := runOnU2FDevicesOnce(unknownHandles, runCredentials); {
+		case err == errKeyMissingOrNotVerified:
```

Comment:
Should we use `errors.Is()` here so that the check succeeds even if the error is wrapped?

Reply:
Done.